core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java

```diff
@@ -12,366 +12,204 @@
    * @param mode the mode to lock in
    */
-  public void lockRootEdge(LockMode mode) {
-    Preconditions.checkState(mEntries.isEmpty());
-
-    mEntries.add(new EdgeEntry(mInodeLockManager.lockEdge(ROOT_EDGE, mode), ROOT_EDGE));
-    mLockMode = mode;
-  }
+  void lockEdge(Inode inode, String childName, LockMode mode);
 
   /**
-   * Leapfrogs the edge write lock forward, reducing the lock list's write-locked scope.
-   *
-   * For example, if the lock list is in write mode with entries [a, a->b, b, b->c],
-   * pushWriteLockedEdge(c, d) will change the list to [a, a->b, b, b->c, c, c->d]. The c->d edge
-   * will be write locked instead of the b->c edge. At least a read lock on b->c will be maintained
-   * throughout the process so that other threads cannot interfere with creates, deletes, or
-   * renames.
+   * Unlocks the last locked inode.
    *
-   * For composite lock lists, this method will do nothing if the base lock is is write locked.
+   * Example
+   * Starting from [a, a->b, b]
    *
-   * @param inode the inode to add to the lock list
-   * @param childName the child name for the edge to add to the lock list
-   */
-  public void pushWriteLockedEdge(Inode inode, String childName) {
-    Preconditions.checkState(!endsInInode());
-    Preconditions.checkState(mLockMode == LockMode.WRITE);
-
-    if (mEntries.isEmpty()) {
-      // Cannot modify the base lock list, and the new inode is already implicitly locked.
-      return;
-    }
-
-    int edgeIndex = mEntries.size() - 1;
-    lockInodeInternal(inode, LockMode.READ);
-    lockEdgeInternal(childName, LockMode.WRITE);
-    downgradeEdge(edgeIndex);
-  }
-
-  /**
-   * @return whether this lock list ends in an inode (as opposed to an edge)
-   */
-  public boolean endsInInode() {
-    return lastEntry() instanceof InodeEntry;
-  }
-
-  /**
-   * Unlocks the last locked inode.
+   * unlockLastInode() results in [a, a->b]
    */
-  public void unlockLastInode() {
-    Preconditions.checkState(endsInInode());
-    Preconditions.checkState(!mEntries.isEmpty());
-
-    mLockedInodes.remove(mLockedInodes.size() - 1);
-    mEntries.remove(mEntries.size() - 1).mLock.close();
-    mLockMode = LockMode.READ;
-  }
+  void unlockLastInode();
 
   /**
    * Unlocks the last locked edge.
+   *
+   * Example
+   * Starting from [a, a->b]
+   *
+   * unlockLastEdge results in [a]
    */
-  public void unlockLastEdge() {
-    Preconditions.checkState(!endsInInode());
-    Preconditions.checkState(!mEntries.isEmpty());
-
-    mEntries.remove(mEntries.size() - 1).mLock.close();
-    mLockMode = LockMode.READ;
-  }
+  void unlockLastEdge();
 
   /**
    * Downgrades the last inode from a write lock to a read lock. The read lock is acquired before
    * releasing the write lock.
+   *
+   * Example
+   * Starting from [a, a->b, b*]
+   *
+   * downgradeLastInode() results in [a, a->b, b]
+   *
+   * If the last inode is not the only write-locked inode, no downgrade occurs.
```

Comment:
If the last inode is **not**(should it be removed?) the only write-locked inode, no downgrade occurs